### PR TITLE
Generate CSRs for kubeadm

### PR DESCRIPTION
**Pull request by liztio**
**What type of PR is this?**

> /kind feature

**What this PR does / why we need it**:
Generating CSRs makes it much easier to integrate kubeadm's certificates into a larger organisation. Previously adminstrators would have to carefully set up certificates in the correct configuration. With a CSR they can simply sign them appropriately and move them to the directory. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #kubernetes/kubeadm#794

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Add option to create CSRs instead of certificates for kubeadm init phase certs and kubeadm alpha certs renew
```

### CODE REVIEW

**Comment by liztio** [comment]
@timothysc that one's included too.

File: cmd/kubeadm/app/cmd/init.go [hunk 1]
```diff
@@ -110,6 +110,8 @@ type initOptions struct {
 	dryRun                bool
 	kubeconfigDir         string
 	kubeconfigPath        string
+	csrOnly               bool
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/init.go]: These don't really belong to init... The option should be plumbed down lower as the option only applies to certs.

File: cmd/kubeadm/app/cmd/init.go [hunk 2]
```diff
@@ -127,10 +129,13 @@ type initData struct {
 	ignorePreflightErrors sets.String
 	certificatesDir       string
 	dryRunDir             string
-	externalCA            bool
-	client                clientset.Interface
-	waiter                apiclient.Waiter
-	outputWriter          io.Writer
+	csrOnly               bool
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/init.go]: Same comment about option order

**Review by timothysc — CHANGES_REQUESTED** [review]
So the biggest issue I see is a layer violation on the init-options. 

Also please squash commits.

**Comment by timothysc** [comment]
/hold 

@fabriziopandini we have a problem on option passing to phases.

File: cmd/kubeadm/app/cmd/init.go [hunk 3]
```diff
@@ -467,6 +485,10 @@ func runInit(i *initData, out io.Writer) error {
 		return errors.Wrap(err, "error getting directories to use")
 	}
 
+	if i.csrOnly {
+		return errors.New("--csr-only can only be used with phase cert")
+	}
+
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/init.go]: the above `if i.csrOnly {` blob is redundant.

File: cmd/kubeadm/app/cmd/init.go (hunk 1 above)
Comment (neolit123) [reply cmd/kubeadm/app/cmd/init.go]: we already it this way for all phases.
my suggestion would be to go with this for 1.13 and think of eventual refactor later.

File: cmd/kubeadm/app/cmd/options/certs.go [hunk 4]
```diff
@@ -22,3 +22,13 @@ import "github.com/spf13/pflag"
 func AddCertificateDirFlag(fs *pflag.FlagSet, certsDir *string) {
 	fs.StringVar(certsDir, CertificatesDir, *certsDir, "The path where to save the certificates")
 }
+
+// AddCSRFlag adds the --csr-only flag to the given flagset
+func AddCSRFlag(fs *pflag.FlagSet, csr *bool) {
+	fs.BoolVar(csr, CSROnly, *csr, "Instead of generating certificates ourselves, create a CSR")
```
Comment (neolit123) [inline cmd/kubeadm/app/cmd/options/certs.go]: `Create CSRs instead of generating certificates`

File: cmd/kubeadm/app/phases/certs/certs.go [hunk 5]
```diff
@@ -276,6 +295,30 @@ func writeKeyFilesIfNotExist(pkiDir string, baseName string, key *rsa.PrivateKey
 	return nil
 }
 
+func writeCSRFilesIfNotExist(csrDir string, baseName string, csr *x509.CertificateRequest, key *rsa.PrivateKey) error {
+	if pkiutil.CSROrKeyExist(csrDir, baseName) {
+		_, _, err := pkiutil.TryLoadCSRAndKeyFromDisk(csrDir, baseName)
+		if err != nil {
+			return errors.Wrapf(err, "%s csr existed but it could not be loaded properly", baseName)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/certs.go]: `csr`- > `CSR`

File: cmd/kubeadm/app/phases/certs/certs.go [hunk 6]
```diff
@@ -276,6 +295,30 @@ func writeKeyFilesIfNotExist(pkiDir string, baseName string, key *rsa.PrivateKey
...
+		}
+
+		fmt.Printf("[certs] Using the existing %q csr\n", baseName)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/certs/certs.go]: `CSR` should be upper case in all errors and stdout output.

File: cmd/kubeadm/app/cmd/init.go (hunk 3 above)
Comment (liztio) [reply cmd/kubeadm/app/cmd/init.go]: This is so you can't use `--csr-only` when you're not calling the `phase` individually. I don't want it running as part of a normal `kubeadm init`

File: staging/src/k8s.io/client-go/util/cert/cert.go [hunk 7]
```diff
@@ -87,6 +87,26 @@ func NewSelfSignedCACert(cfg Config, key crypto.Signer) (*x509.Certificate, erro
 	return x509.ParseCertificate(certDERBytes)
 }
 
+// NewCSR creates a new CSR
+func NewCSR(cfg Config, key crypto.Signer) (*x509.CertificateRequest, error) {
```
Comment (mikedanese) [inline staging/src/k8s.io/client-go/util/cert/cert.go]: the certificates packages in client-go are kind of a kitchen sink and we are working on removing a lot of the surface. @awly would you keep these functions here? I'm wondering if adding this to kubeadm utilities would be preferable.

Comment (awly) [reply staging/src/k8s.io/client-go/util/cert/cert.go]: I'd move this to pkiutil.
Everything under client-go is intended for clients of k8s that do not want to vendor the entire k8s codebase. We should keep only things that are relevant for clients here.

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 8]
```diff
@@ -112,6 +127,20 @@ func WriteKey(pkiPath, name string, key *rsa.PrivateKey) error {
...
 
+// WriteCSR stores the given CSR at the given location
+func WriteCSR(csrDir, name string, csr *x509.CertificateRequest) error {
+	if csr == nil {
+		return errors.New("certificate request cannot be nil when writing to file")
+	}
+
+	csrPath := pathForCSR(csrDir, name)
+	if err := certutil.WriteCSR(csrPath, certutil.EncodeCSRPEM(csr)); err != nil {
+		return errors.Wrapf(err, "unable to write csr to file %s", csrPath)
```
Comment (atoato88) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: Is it correct that csr -> CSR ?

File: cmd/kubeadm/app/phases/certs/certs_test.go [hunk 9]
```diff
@@ -606,6 +690,38 @@ func TestValidateMethods(t *testing.T) {
...
+
+	certConfig, err := kubeadmCert.GetConfig(cfg)
+	if err != nil {
+		t.Fatalf("couldn't get cert config: %v", err)
+	}
+
+	csr, _, err := NewCSR(&kubeadmCert, cfg)
+
+	if err != nil {
+		t.Errorf("invalid signature on csr: %v", err)
```
Comment (atoato88) [inline cmd/kubeadm/app/phases/certs/certs_test.go]: Is it correct that csr -> CSR ?

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 10]
```diff
@@ -65,6 +65,21 @@ func NewCertAndKey(caCert *x509.Certificate, caKey *rsa.PrivateKey, config *cert
 	return cert, key, nil
 }
 
+// NewCSRAndKey generates a new CSR and key that could be signed to create the given
```
Comment (rosti) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: Unfinished comment?

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 11]
```diff
@@ -65,6 +65,21 @@ func NewCertAndKey(caCert *x509.Certificate, caKey *rsa.PrivateKey, config *cert
...
+func NewCSRAndKey(config *certutil.Config) (*x509.CertificateRequest, *rsa.PrivateKey, error) {
+	key, err := certutil.NewPrivateKey()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "unable to create private key")
+	}
+	csr, err := certutil.NewCSR(*config, key)
+
```
Comment (rosti) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: Strange NL choice here. Probably move the empty line before the `NewCSR` call?

File: cmd/kubeadm/app/phases/certs/certs.go [hunk 12]
```diff
@@ -276,6 +295,30 @@ func writeKeyFilesIfNotExist(pkiDir string, baseName string, key *rsa.PrivateKey
 	return nil
 }
 
+func writeCSRFilesIfNotExist(csrDir string, baseName string, csr *x509.CertificateRequest, key *rsa.PrivateKey) error {
```
Comment (rosti) [inline cmd/kubeadm/app/phases/certs/certs.go]: I'd rather put a big comment on top of this one, to prevent people from accidentally calling it.

File: cmd/kubeadm/app/phases/certs/certs.go [hunk 13]
```diff
@@ -276,6 +295,30 @@ func writeKeyFilesIfNotExist(pkiDir string, baseName string, key *rsa.PrivateKey
...
+	if pkiutil.CSROrKeyExist(csrDir, baseName) {
+		_, _, err := pkiutil.TryLoadCSRAndKeyFromDisk(csrDir, baseName)
```
Comment (rosti) [inline cmd/kubeadm/app/phases/certs/certs.go]: Do we really have to load the key here? In the wake of Meltdown and Spectre I am a bit hesitant to load a key in memory, just to check if it exists. Probably a good old `os.Stat` of a `*.key` file can do the trick here?
It's better to keep it in this way in the testing code though (but not in production code).

**Review by rosti — COMMENTED** [review]
Thanks @liztio !
Only some tiny nits on my behalf.

File: cmd/kubeadm/app/cmd/phases/certs.go [hunk 14]
```diff
@@ -47,6 +48,11 @@ var (
 		` + cmdutil.AlphaDisclaimer)
 )
 
+var (
+	csrOnly bool
```
Comment (liztio) [inline cmd/kubeadm/app/cmd/phases/certs.go]: I don't love that we have to do this, but I don't see any other way to plum this

File: staging/src/k8s.io/client-go/util/cert/cert.go (hunk 7 above)
Comment (liztio) [reply staging/src/k8s.io/client-go/util/cert/cert.go]: Ack. Everything removed from staging.

File: cmd/kubeadm/app/phases/certs/certs.go (hunk 13 above)
Comment (liztio) [reply cmd/kubeadm/app/phases/certs/certs.go]: I'd prefer to keep this consistent with the other similar methods. Maybe file an issue against kubernetes/kubeadm to fix this everywhere? It should be pretty straightforward

**Comment by timothysc** [comment]
It looks like this needs a rebase now that @fabriziopandini 's PR has merged. 
/hold cancel

File: cmd/kubeadm/app/cmd/init.go [hunk 15]
```diff
@@ -117,10 +117,11 @@ type initData struct {
 	ignorePreflightErrors sets.String
 	certificatesDir       string
 	dryRunDir             string
-	externalCA            bool
-	client                clientset.Interface
-	waiter                apiclient.Waiter
-	outputWriter          io.Writer
+
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/init.go]: There is no change here that relates to your PR.

File: cmd/kubeadm/app/cmd/init.go [hunk 16]
```diff
@@ -236,23 +237,23 @@ func AddInitConfigFlags(flagSet *flag.FlagSet, cfg *kubeadmapiv1beta1.InitConfig
 }
 
 // AddInitOtherFlags adds init flags that are not bound to a configuration file to the given flagset
-func AddInitOtherFlags(flagSet *flag.FlagSet, cfgPath *string, skipTokenPrint, dryRun *bool, ignorePreflightErrors *[]string) {
+func AddInitOtherFlags(flagSet *flag.FlagSet, initOpts *initOptions) {
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/init.go]: I'm ok with this, but this seems orthogonal to the purpose of the PR.

File: cmd/kubeadm/app/cmd/phases/certs_test.go [hunk 17]
```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2017 The Kubernetes Authors.
```
Comment (timothysc) [inline cmd/kubeadm/app/cmd/phases/certs_test.go]: 2018

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 18]
```diff
@@ -112,6 +132,28 @@ func WriteKey(pkiPath, name string, key *rsa.PrivateKey) error {
...
+// If the CSR file already exists, it will be overwritten.
+// The parent directory of the csrPath will be created as needed with file mode 0755.
+func WriteCSR(csrDir, name string, csr *x509.CertificateRequest) error {
+	if csr == nil {
+		return errors.New("certificate request cannot be nil when writing to file")
+	}
+
+	csrPath := pathForCSR(csrDir, name)
+	if err := os.MkdirAll(filepath.Dir(csrPath), os.FileMode(0755)); err != nil {
+			return err
```
Comment (timothysc) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: Wrap for consistency, also to denote location for debugging.

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 19]
```diff
@@ -112,6 +132,28 @@ func WriteKey(pkiPath, name string, key *rsa.PrivateKey) error {
...
+	}
+
+
```
Comment (timothysc) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: nix spare line

File: cmd/kubeadm/app/util/pkiutil/pki_helpers.go [hunk 20]
```diff
@@ -207,6 +260,23 @@ func TryLoadKeyFromDisk(pkiPath, name string) (*rsa.PrivateKey, error) {
...
+	csrPath := pathForCSR(pkiPath, name)
+
+	csr, err := CertificateRequestFromFile(csrPath)
+	if err != nil {
+		return nil, nil, errors.Wrapf(err, "couldn't load the certificate request %s", csrPath)
+	}
+
+	key, err := TryLoadKeyFromDisk(pkiPath, name)
+	if err != nil {
+		return nil, nil, err
```
Comment (timothysc) [inline cmd/kubeadm/app/util/pkiutil/pki_helpers.go]: same re: Wrap.

**Review by timothysc — DISMISSED** [review]
Minor comments.

**Review by neolit123 — APPROVED** [review]
LGTM
thanks for all the updates @liztio . i don't see anything blocking.
